sd-core/src/language/llvm_ir.rs

```diff
@@ -0,0 +1,670 @@
+use derive_more::{Display, From};
+use either::Either;
+use itertools::Itertools;
+use llvm_ir::{
+    instruction::{InlineAssembly, RMWBinOp},
+    module::{GlobalAlias, GlobalIFunc, GlobalVariable},
+    BasicBlock, Constant, ConstantRef, FPPredicate, Function, Instruction, IntPredicate, Module,
+    Name, Operand, Terminator,
+};
+use tracing::debug;
+
+use super::{Fresh, Language, OpInfo, CF};
+use crate::{
+    common::Matchable,
+    hypergraph::traits::{WireType, WithType},
+};
+
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
+pub struct LlvmIrSettings {
+    pub sym_name_linking: bool,
+}
+
+impl Default for LlvmIrSettings {
+    fn default() -> Self {
+        Self {
+            sym_name_linking: true,
+        }
+    }
+}
+
+pub struct LlvmIr;
+
+pub type Expr = super::Expr<LlvmIr>;
+pub type Bind = super::Bind<LlvmIr>;
+pub type Value = super::Value<LlvmIr>;
+pub type Thunk = super::Thunk<LlvmIr>;
+pub type Block = super::Block<LlvmIr>;
+
+impl Language for LlvmIr {
+    type Op = Op;
+    type Var = Var;
+    type Addr = String;
+    type VarDef = Var;
+    type BlockAddr = Name;
+    type Symbol = Box<String>;
+}
+
+#[derive(Clone, PartialEq, Debug, Hash, Display, From)]
+pub enum Op {
+    // constant-like
+    Constant(ConstantRef),
+    RMWBinOp(RMWBinOp),
+    IntPredicate(IntPredicate),
+    FPPredicate(FPPredicate),
+    // globals
+    #[display("inline assembly")]
+    InlineAssembly(InlineAssembly),
+    #[display("{}", _0.name)]
+    GlobalVariable(GlobalVariable),
+    #[display("{}", _0.name)]
+    GlobalAlias(GlobalAlias),
+    #[display("{}", _0.name)]
+    GlobalIFunc(GlobalIFunc),
+    // end of basic block
+    Terminator(Terminator),
+    // LLVM machine instructions
+    Instruction(Instruction),
+}
+
+impl Eq for Op {}
```

Comment:
Is this not derivable?

Reply:
No, because it requires an `Eq` bound on all the enum variants, extending into `llvm-ir` (there, the relevant types are only `PartialEq` - not really sure why they aren't also `Eq`). We had been using `derivative` to get around this.